### ISSUE TITLE
Performance investigation, RAM usage and paging

### ISSUE THREAD

**Issue by johndoe31415** [OP-TEE/optee_os#5915]
I'm developing on the STM32MP1 platform, specifically the STM32MP157 using embedded Linux and OP-TEE. As a short disclaimer, I'm using the STM fork of OP-TEE with the currently supported version 3.16, on my target but for the following comments I've only used vanilla OP-TEE (in version 3.16, however).

Functionally, in my case, everything I care about (trusted keys and PKCS#11) works. There's a catch, however: especially PKCS#11 handshakes in OP-TEE are horrifically slow. Even with TLS session tickets and session resumption, an initial handshake using a P-256 certificate takes a whooping 5 seconds. Using the exact same cryptographic primitivies and no OP-TEE establishes that handshake in 100ms. This may be fully prohibitive because timeouts on certain AWS services are not configurable and hence it is *impossible* to successfully establish a connection there. Even if it was, 5 seconds is simply a very long time.

So I am on the journey to answer two questions primarily:

1. Why is performance so bad and what are ways to improve it?
2. How close "to the edge" am I?

(1) is fairly straightforward, while (2) may be more nuanced. Here's what I mean by it: Let's say I get it to run in acceptable time (500ms) with some tricks. Then, possibly due to minor changes in the code or compiler, a future version could be pushed above a threshold of RAM usage and push me back into the "dead zone" again. I know this is an issue becuase, for example, when I turn on logging, the binaries become so large that my STM32 doesn't boot anymore. This is making me uncomfortable and I would like to have a safe, quantifiable margin of "buffer" before pushing forward.

In any case, my initial suspicion was, because I have previously had severe problems with too large images, that OP-TEE was heavily paging and that complely destroys performance. So what I did was an investigation in QEMU only for starters. I have a script that can instrument both the debugging interface/gdb as well as the console and automatically start QEMU instances. So I built OP-TEE 3.16 a total of 64 times with configured SRAM sizes (`CFG_CORE_TZSRAM_EMUL_SIZE`) of 256 kiB to 512 kiB in 4 kiB increments. Side note: The STM32MP1 has 256 kiB of SYSRAM, which I guess should be enough (the documentation says "But those with 128 to 256 KiB of SRAM can be expected to have a capable TEE solution in SRAM.").

With each image, I initialized the PKCS#11 interface and measured the time to create a P-256 key. I know that key generation is not indicative of handshake performance (we actually only need ECDSA signing in OP-TEE, ephemeral X25519 key agreement can be done outside OP-TEE), but it gives an interesting data point. I'll follow up with more measurements soon. In either case I created two keys just after each other (to counteract pages that would possibly need to be paged in the first run). Long story short, here are the results:

![profile_keygen](https://user-images.githubusercontent.com/5919339/227803615-ee7318ed-c71c-430e-b02a-2ee6d089331e.png)

Here is the image header:

```
Version     : 2
Architecture: AArch32
Flags       : 0
Images      : 2

Image 0:
    Load address: 0xe100000
    Image ID    : Pager
    Size        : 93992 = 0x16f28 = 91 kiB + 808 B

Image 1:
    Load address: 0xffffffffffffffff
    Image ID    : Paged
    Size        : 401408 = 0x62000 = 392 kiB
```

This is the configuration I'm using to build:

```json
{
		"BR2_PACKAGE_OPENSC": "y",
		"BR2_PACKAGE_LIBOPENSSL": "y",
		"BR2_PACKAGE_LIBOPENSSL_BIN": "y",
		"BR2_PACKAGE_LIBP11": "y",
		"CFG_CORE_ASLR": "n",
		"CFG_CORE_LOG": "1",
		"CFG_TEE_CORE_LOG_LEVEL": "1",
		"CFG_TEE_TA_LOG_LEVEL": "1",
		"CFG_TEE_CORE_DEBUG": "n",
		"CFG_WITH_PAGER": "y",
		"CFG_CORE_TZSRAM_EMUL_SIZE": "<variable>"
}
```

Interesting are that there are two distinct "steps", probably where the number of pages is sufficient to fully contain a particular chunk of code. At roughly 91 + 392 kiB = 483 kiB no improvement should be expected anymore (everything is fully paged in) and that's also what the numbers indicate.

The code I'm using is here: https://github.com/johndoe31415/opteetk

Note that I believe that this issue: https://github.com/OP-TEE/optee_os/issues/3579 is possibly seeing the same phaenomenon (I was unable to get a hold of the HiKey 960 datasheet to confirm how much SRAM that SoC has and the publicly available information doesn't mention it).

Is this a known issue? Are 256 kiB of SRAM too little to have good performance using P-256 or is this something to further investigate? Are there any suggested remediations to get code smaller that I could try?

My next steps are to profile code (on QEMU again) using actual TLS handshakes and put the numbers here. I've also written an experimental stub to gdb which is able to automatically handle page faults and read out the page directory at every fault so that I can cross-check with the ELF and validate if this is fully expected or if there is potential for optimization (paging strategy, memory layout). I'll keep you posted.

**Comment by johndoe31415** [comment]
So I've made the TLS handshake timing measurements and they are similar in nature to the key generation. Measured was a client certificate TLS handshake using P-256 certificates. They are really bad for the 256 kiB case (9 sec per handshake, on the actual target I get 5 sec), but rapidly improve. The sweet spot seems to be somewhere around 384 kiB with asymptotic performance of about 1 sec per handshake in QEMU. For comparison, software TLS handshake on QEMU on my computer is around 200ms, so we have a performance penalty of factor 5 for OP-TEE (not great, but definitely doable -- if it's somewhat linear that would mean 500ms on the target, which would be a 10-fold improvement over what I'm seeing now). Curious what you think about this and if this is expected behavior (it does seem like quite a high amount of required RAM IMHO).

![profile_tls_handshake](https://user-images.githubusercontent.com/5919339/227809760-e9e3a3e0-f4b9-4d9f-814b-6379a39c5c1c.png)

**Comment by jenswi-linaro** [comment]
For the paging problem, perhaps it's possible to achieve better locality in the code so fewer pages are involved. However, I'm not sure how to do that and it may very well result in a horrible hack.

Signing and key generation have always felt a bit slower than expected. People have in the past tried to investigate why this is, but we have never found an answer.

By the way, I like your charts. :-)

**Comment by johndoe31415** [comment]
> Hi @johndoe31415, Indeed STM32MP15 256kB internal memory is quite small. Pager efficiency depends on number of pages available in the pager pool and on how many pages need to be swapped for executtion of the related OP-TEE service. Instead of monitoring SRAM size, look more specifically at the pager pool size. The smaller the pool size, the more swapped pages. From experience, crypto operations are quite smooth with a pool above of 128kB. 

Interesting. For my experiments I have tried to have a QEMU build as closely as possible to the STM32MP1 target I'm developing on and hence have logging disabled, but in my example, `xtest --stats --pager` shows a page pool size of 19 pages, which would be only 76 kiB. Does 180 kiB of required, statically allocated memory seem a bit much? It feels that way, but I have no good data to go on.

> Below this size some perfs starts to degrade, and there are thresholds where degratation is dramatic. Pager pool uses all SRAM that is not resisdent memory. There is an info level trace on pager pool size, e.g. `I/TC: Pager pool size: 104kB` `xtest --stats --pager` shows some pager numbers. Note on the stm32mp15/optee firmware you use, there are pager resident code for the low power sequences that are executed on PSCI fastcall SMC calls and consume resident memory to the detriment of the pager pool size. Enabling some debug features do consume quite some resident RAM (`CFG_TEE_CORE_DEBUG=y` > 24kB, `CFG_UNWIND=y` ~20kB, `CFG_TEE_CORE_LOG_LEVEL` level 1 and 2 => ~8kB, log level 3 => ~12kB, ...) .

Also great hints. I tried to debug where pagefaults occur during a TLS handshake using ECC in OP-TEE. For this, as mentioned before, I used a custom gdb Python script which inserted a breakpoint on `tee_pager_handle_fault` and, on each occurence (page preemtion event) it dumped the whole `tee_pager_pmem_head` linked listed into a JSON file. This resulted in the words slowest TLS handshake (took about 1.5 hours to complete, whooping ~300000 page faults!)

During postprocessing, I then examined that list. I'm not exactly certain I understand, but what I did is compute the address of the page that was replaced by looking at the last entry of the list, then took `fobj` (the page VMA, as I understand it) and added 4 kiB * `fobj_pgidx`. This gave me very interesting results. In postprocessing, I then looked at the symbols and symbol sizes of `tee.elf`, reconciled with the affected pages and hence have a list of page fault counts by their page address and the symbols which lie within that page. The data is super interesting. Here's a peek:

```
 0xe116000  27371 main_mmu_l1_ttb
 0xe117000  27369 main_mmu_l1_ttb
 0xe17a000  27177 __tee_alg_get_class __tee_alg_get_digest_hash __tee_alg_get_main_alg __tee_alg_hash_algo get_attribute get_user_u64_as_size_t op_attr_bignum_clear op_attr_bignum_free op_attr_bignum_from_binary op_attr_bignum_from_obj op_attr_bignum_from_user op_attr_bignum_to_binary op_attr_bignum_to_user op_attr_secret_value_clear op_attr_secret_value_from_binary op_attr_secret_value_from_obj op_attr_secret_value_from_user op_attr_secret_value_to_binary op_attr_secret_value_to_user op_attr_value_clear op_attr_value_from_binary op_attr_value_from_obj op_attr_value_from_user op_attr_value_to_binary op_attr_value_to_user op_u32_from_binary_helper op_u32_to_binary_helper put_user_u64 set_attribute syscall_cryp_obj_get_attr syscall_cryp_obj_get_info syscall_cryp_obj_restrict_usage syscall_get_time syscall_mask_cancellation syscall_set_ta_time syscall_unmask_cancellation syscall_wait tee_obj_attr_free tee_svc_cryp_obj_find_type_attr_idx tee_svc_find_type_props
 0xe174000  27147 do_range do_range_backwards htree_test_rewrite open_session read_block system_dlopen system_dlsym system_get_tpm_event_log system_supp_plugin_invoke test_get_offs_size test_read_final test_read_init test_write_final test_write_init uint_to_ptr val_from_bn_n_salt write_block
 0xe173000  27009 __tee_alg_get_digest_hash __tee_alg_hash_algo add_ta bootstrap check_install_conflict get_alloc_stats get_devices get_memleak_stats get_pager_stats install_ta invoke_command mobj_put shdr_free shdr_get_size system_derive_ta_unique_key system_map_zi system_rng_reseed system_unmap tee_mmu_user_va2pa_attr tpm_get_event_log vm_check_access_rights vm_set_ctx
 0xe153000  26195 _ltc_ecc_generate_keypair _ltc_ecc_shared_secret _ltc_ecc_sign _ltc_ecc_verify crypto_acipher_alloc_dsa_keypair crypto_acipher_alloc_dsa_public_key crypto_acipher_alloc_rsa_keypair crypto_acipher_alloc_rsa_public_key crypto_acipher_dsa_sign crypto_acipher_dsa_verify crypto_acipher_gen_dsa_key crypto_acipher_gen_rsa_key crypto_acipher_rsaes_decrypt crypto_aes_ccm_alloc_ctx crypto_aes_ccm_dec_final crypto_aes_ccm_enc_final crypto_aes_ccm_init crypto_aes_ccm_update_aad crypto_aes_ccm_update_payload crypto_aes_ctr_alloc_ctx crypto_aes_xts_alloc_ctx crypto_asym_alloc_ecc_keypair crypto_asym_alloc_ecc_public_key crypto_des3_cbc_alloc_ctx crypto_des_cbc_alloc_ctx ecc_get_curve_info ecc_populate_ltc_private_key ecc_populate_ltc_public_key ecc_set_curve_from_name ltc_ctr_init ltc_ctr_update ltc_xts_init ltc_xts_update rsadorep tee_algo_to_ltc_hashindex
 0xe154000  25192 crypto_acipher_alloc_dh_keypair crypto_acipher_dh_shared_secret crypto_acipher_gen_dh_key crypto_acipher_rsaes_encrypt crypto_acipher_rsassa_sign crypto_acipher_rsassa_verify crypto_aes_expand_enc_key sm2_ltc_dsa_sign sm2_ltc_dsa_verify sm2_ltc_pke_decrypt sm2_ltc_pke_encrypt
 0xe177000  24776 bm_timestamp cleanup_shm_refs copy_out_param core_mutex_tests core_self_tests do_update entry_close_session entry_invoke_command entry_open_session free_ctx get_open_session_meta init_ctx mobj_get mobj_put mutex_test_reader mutex_test_writer param_mem_from_mobj self_test_nex_malloc set_rmem_param set_tmem_param update_ae update_cipher
 0xe175000  23918 aux_alloc aux_free core_fs_htree_tests core_lockdep_tests test_corrupt test_corrupt_type test_entry_memref_null test_entry_params test_trace test_v2p2v test_write_read
 0xe179000  21085 alloc_temp_sec_mem get_prop_client_id get_prop_set get_prop_struct get_prop_ta_app_id get_prop_tee_sys_time_prot_level syscall_check_access_rights syscall_close_ta_session syscall_get_cancellation_flag syscall_get_property syscall_get_property_name_to_index syscall_invoke_ta_command syscall_open_ta_session tee_svc_copy_param tee_svc_update_out_param
 0xe171000  20453 alloc_pgt cmp_region_for_remap find_vm_region get_num_req_pgts merge_vm_range rem_um_region split_vm_range split_vm_region umap_add_region va_range_is_contiguous vm_map_pad vm_remap
 0xe176000  18482 close_session create_ta destroy_ta open_session self_test_add_overflow self_test_division self_test_malloc self_test_mul_signed_overflow self_test_mul_unsigned_overflow self_test_sub_overflow test_dump_sdp test_inject_sdp test_transform_sdp
 0xe141000  11691 ts_bin_f04a0fe71f5d4b9babf7619b85b4ce8c
 0xe178000   6160 __tee_alg_hash_algo default_mobj_init entry_cancel get_prop_tee_dev_id hkdf_expand hkdf_extract mobj_get_va mobj_put mobj_put_wipe nsec_sessions_list_head pbkdf2_f plat_rng_init register_shm syscall_log syscall_not_supported tee_alg_get_digest_size tee_cryp_concat_kdf tee_cryp_hkdf tee_cryp_init tee_cryp_pbkdf2 tee_do_cipher_update tee_entry_std tee_hash_createdigest tee_obj_add tee_obj_alloc tee_obj_close tee_obj_close_all tee_obj_free tee_obj_get unregister_shm
 0xe13e000   1128 ts_bin_f04a0fe71f5d4b9babf7619b85b4ce8c
 0xe140000    700 ts_bin_f04a0fe71f5d4b9babf7619b85b4ce8c
 0xe13c000    566 CK SboxTable __func__.5 __func__.6 aes_gcm_ops aes_key aes_key2 api_vers arm_cntpct_time_source crypto_cbc_mac_ops cts_ops descr fw_impl_bin_version fw_impl_version fw_manufacturer ldelf_syscall_table pattern.1 pl011_driver pl011_match_table pseudo_ta_ops simple_bus_match_table sm3_hash_ops sm3_hmac_ops sm3_padding sm4_cbc_ops sm4_ctr_ops sm4_ecb_ops ta_time_prot_lvl ta_ver_db_obj_id tee_svc_syscall_table test_htree_ops trustedos_impl_bin_version trustedos_impl_version trustedos_manufacturer ts_antiroll_prot_lvl vendor_props_tee vendor_propset_array_tee zero_salt
 0xe146000    304 ts_bin_f04a0fe71f5d4b9babf7619b85b4ce8c
 0xe14a000    290 Rorder Worder crypto_ecc_en der_asn1_tag_to_type_map der_asn1_tag_to_type_map_sz der_asn1_type_to_identifier_map der_asn1_type_to_identifier_map_sz ia5_table pc1 pc2 printable_table small_prime totrot
 0xe12c000    267 _pbuf_intersects _pbuf_is_inside add_phys_mem add_va_space asan_memcpy_unchecked boot_init_primary_early carve_out_asan_mem find_map_by_pa find_map_by_type fobj_put init_asan init_canaries init_primary init_runtime init_thread_stacks init_user_kcode init_vcore mutex_lockdep_init pbuf_inside_map_area pbuf_is_nsec_ddr pbuf_is_sdp_mem pbuf_is_special_mem primary_save_cntfrq print_pager_pool_size print_stack_limits ro_paged_alloc tee_mm_get_size thread_init_boot_thread thread_init_primary thread_init_stack thread_init_threads undo_init_relocation
 0xe145000    228 ts_bin_f04a0fe71f5d4b9babf7619b85b4ce8c
 0xe148000    219 SP1 SP2 SP3 SP4 SP5 SP6 SP7 SP8 TD0 Td4 bigbyte bytebit des3_desc des_desc prng_crypto_desc sm2_pke_public_key_ops
 0xe115000    186 main_mmu_l1_ttb
 0xe147000    186 aes_ccm_ops dbase.1 dext.0 distfix.1 ecc_keypair_ops ecc_public_key_ops lbase.3 lenfix.2 lext.2 ltc_cbc_ops ltc_ctr_ops ltc_ecb_ops ltc_hash_ops ltc_hmac_ops ltc_omac_ops ltc_xts_ops order.0 sm2_dsa_keypair_ops sm2_dsa_public_key_ops sm2_kep_keypair_ops sm2_kep_public_key_ops sm2_pke_keypair_ops
 0xe15f000    186 abort_print_current_ts cpu_to_fdt32 cpu_to_fdt64 fdt32_to_cpu fdt64_to_cpu fdt_setprop_cell fdt_setprop_u32 get_dt get_rpc_alloc_res get_rpc_arg get_rpc_arg_res rpc_shm_mobj_alloc set_rmem set_tmem thread_rpc_alloc thread_rpc_alloc_arg thread_rpc_alloc_global_payload thread_rpc_alloc_kernel_payload thread_rpc_alloc_payload thread_rpc_cmd thread_rpc_free thread_rpc_free_global_payload thread_rpc_free_kernel_payload thread_rpc_free_payload tpm_map_log_area
```

A few things interesting about this:

- It seems the page directory itself gets preemted and needs to be swapped in *very* frequently. It would be interesting to experiment with locking specific memory addresses and never preempt them. Sacrifice a single page or two (actually `main_mmu_l1_ttb` is 4 pages large, 0x4000 bytes) but gain a lot less preemtion. My guess is that the page fault handler may need to access the TTB and hence cause a followup page fault?
- The top 19 page fault addresses (that's the size of my paging pool) make for 317013 of 320987 recorded page faults, or 98.8%. So if we somehow magically knew the 18 "hottest" pages and pinned them all before a specific operation and only used the single remaining page, I believe performance could be dramatically improved in low-RAM scenarios. What's your thought on this?

**Comment by johndoe31415** [comment]
Some more data that is puzzling me right now. So with the data I collected I wrote a userland script to simulate different preemption strategies (e.g., usage counter, etc) to get a feeling for how they would behave with actual data. The first strategy I've implemented is LRU, of which I assumed this is what OP-TEE is using right now. It's definitely the easiest to implement.

Now it gets weird: cross-checking with LRU should give me the same ballpark of page faults as the actual implementation (unless I'm doing something wrong). But it isn't. Simulating with simple LRU gives me only 2180 page faults. I'm seeing about 321000 in reality and really cannot explain why at all.

Here's the tool I wrote: https://github.com/johndoe31415/opteetk/blob/main/simulate_pagefault_strategy (includes also the gdb stub I'm using to extract the data). If you're interested in the data I've collected, this is here: https://sponkuller.de/tls-handshake-256kib-sram.txt.gz

You just have to gunzip it and can run it:

```
$ ./simulate_pagefault_strategy tls-handshake-256kib-sram.txt 
Page faults: 2180
```

**Comment by etienne-lms** [comment]
About top pages that are being faulted, pager unmaps page that are valid so that they produce page fault quickly sorted out (no physical page swapping needed) hence get listed as pages often used so that they preferrably not paged out.
If tracinkg page fualt, the interresting are those that lead to a page swap (eviction and load).

**Comment by johndoe31415** [comment]
@etienne-lms Ah, yes of course, I assumed that the pagefault handler would always evict, but didn't consider that you're sometimes just changing visibility/permissions. However, I don't seem to locate the smoking gun where eviction/loading happens. I had assumed that memory is paged out into insecure RAM and hence wanted to follow a crypto trail. Indeed I found `tee_pager_generate_authenc_key`, but it is only referenced in `tee_pager.h`, nowhere else. This is a bit confusing. Could you please point me in the right direction?

I still want to generate a full memory trace, but this is surprisingly difficult:

- In gdb, I can't seem to properly get hardware breakpoints working with large memory regions (required to capture accurate data not only from code executions but also data loads)
- In QEMU I used the tracing functionality and recorded the memory trace. The data is incomplete, however, since trace events are only generated when the memory fastpass is not used. Fastpass is used in the majority of cases, however.
- Single stepping in gdb would work, but then I would need to manually inspect each instruction to see if it performs memory indirect access, then resolve this manually. A lot of work and would have horrific performance.
- Currently looking into Cannoli: https://github.com/MarginResearch/cannoli which looks very promising but unfortunately the QEMU build I've done that supports Cannoli doesn't (yet?) work to emulate OP-TEE. I'll need to look into the build in more detail.

**Comment by etienne-lms** [comment]
Page eviction and loading sequences in pager are done from calls to `fobj_save_pager()` and `fobj_load_page()` done from `pager_ger_page()` and `pager_deploy_page()`. The sequence to save/load a page depends on the file-objet related to the assigned page (ro, rw, locked, reloc, iv, ...): see `struct fobj_ops` operation handlers from _core/mm/fobj.c_.

**Comment by johndoe31415** [comment]
I've taken another crack at trying to profile this, but it is surprisingly difficult. My first idea was to get cannoli, a qemu memory tracer running to be able to accurately export all page accesses (code and data). When I finally did get cannoli running, I realized that it is only implementing QEMU user mode, not softmmu. Hence for tracing OP-TEE, it is useless. So I tried patching QEMU myself. This isn't straightforward since QEMU compiles code via JIT using the TCG (tiny code generator) engine and, honestly, I went down that rabbit hole far too deep and really didn't get anything useful working -- I could trace some accesses, but not all of them (not reliably) and I believe the reason lies within my lack of understanding of various ARM MMU tables (different directory tables are used in different contexts and I'm not sure how this is handled in QEMU).

Anyways. I was not prepared to give up just yet. So my idea was to build a QEMU variant with extremely little memory to force eviction all the time, then attach gdb to QEMU, break on the eviction save/load functions and trace those. Before doing that I replaced the AES-GCM save/load code of rwp pages by simple memcpy for performance reasons. Still, it didn't really work. First, I had the pager size reduced to only very few pages (1 or 2) and that stopped QEMU from even booting (hung in the beginning). With around 8 pages it actually came up. I had it runnign for close to 24 hrs, millions of evictions, but no meaningful data. Very frustrating. Obviously, I could use `printf()` in the eviction functions, but that would change the memory behavior of the code which is exactly what I do not want.

In a last attempt, I tried attaching gdb to the QEMU process itself and tried using hardware breakpoints to monitor accesses to the memory that is allocated inside QEMU for its guest. Again a rabbit hole I dug in far too deep and, in the end, it did not work.

So all in all this was quite disappointing. Really no smoking gun, no real result other than "OP-TEE kind of seems to need upwards of 256 kB". I was so much hoping there would be a tangible outcome or something easy to change to get it to work with better performance.

For now, I'm focusing my efforts on getting more RAM to OP-TEE. I have a few ideas but these require

1. a closer understanding of the ARM MMU (another rabbit hole I'm in right now)
2. understanding of the MMU programming interaction between TF-A and OP-TEE (who does what?)
3. understanding of the TZ isolation mechanisms used on various peripherals of the STM32MP1

Since all of these three topics are related and not really the performance investigation anymore, I'll close this ticked and open another one where I discuss what I'm attempting in more detail. Thanks everyone for the great help over here!
